cf-check/dump.c

```diff
@@ -0,0 +1,207 @@
+#include <dump.h>
+
+#ifdef LMDB
+#include <lmdb.h>
+#include <string_lib.h>
+
+typedef enum
+{
+    DUMP_MODE_DEFAULT,
+    DUMP_MODE_KEYS,
+    DUMP_MODE_VALUES,
+} dump_mode;
+
+static void print_usage(void)
+{
+    printf("Usage: cf-check dump FILE\n");
+    printf("Example: cf-check dump /var/cfengine/state/cf_lastseen.lmdb\n");
+}
+
+static int dump_report_error(int rc)
+{
+    printf("err(%d): %s\n", rc, mdb_strerror(rc));
+    return rc;
+}
+
+void dump_print_json_string(const char *const data, const size_t size)
+{
+    printf("\"");
+    for (size_t i = 0; i < size; ++i)
+    {
+        char byte = data[i];
+        switch (byte)
+        {
+        case '\0':
+            printf("\\0");
+            break;
+        case '\\':
+            printf("\\\\");
+            break;
+        case '\b':
+            printf("\\b");
+            break;
+        case '\f':
+            printf("\\f");
+            break;
+        case '\n':
+            printf("\\n");
+            break;
+        case '\r':
+            printf("\\r");
+            break;
+        case '\t':
+            printf("\\t");
+            break;
+        case '"':
+            printf("\\\"");
+            break;
+        default:
+            if (isprint(byte))
+            {
+                printf("%c", byte);
+            }
+            else
+            {
+                printf("\\x%2.2x", (unsigned char) byte);
+            }
+        }
+    }
+    printf("\"");
+}
+
+void dump_print_opening_bracket(const dump_mode mode)
+{
+    if (mode == DUMP_MODE_DEFAULT)
+    {
+        printf("{\n");
+    }
+    else
+    {
+        printf("[\n");
+    }
+}
+
+void dump_print_object_line(const MDB_val key, const MDB_val value)
+{
+    printf("\t");
+    dump_print_json_string(key.mv_data, key.mv_size);
+    printf(": ");
+    dump_print_json_string(value.mv_data, value.mv_size);
+    printf(",\n");
+}
+
+void dump_print_array_line(const MDB_val value)
+{
+    printf("\t");
+    dump_print_json_string(value.mv_data, value.mv_size);
+    printf(",\n");
+}
+
+void dump_print_closing_bracket(const dump_mode mode)
+{
+    if (mode == DUMP_MODE_DEFAULT)
+    {
+        printf("}\n");
+    }
+    else
+    {
+        printf("]\n");
+    }
+}
+
+int dump_db(const char *file, const dump_mode mode)
+{
+    assert(file != NULL);
+
+    int r;
+    MDB_env *env;
+    MDB_txn *txn;
+    MDB_dbi dbi;
+    MDB_cursor *cursor;
+
+    if (0 != (r = mdb_env_create(&env))
+        || 0 != (r = mdb_env_open(env, file, MDB_NOSUBDIR | MDB_RDONLY, 0644))
+        || 0 != (r = mdb_txn_begin(env, NULL, MDB_RDONLY, &txn))
+        || 0 != (r = mdb_open(txn, NULL, 0, &dbi))
+        || 0 != (r = mdb_cursor_open(txn, dbi, &cursor)))
+    {
+        return dump_report_error(r);
+    }
+
+    MDB_val key, value;
+    dump_print_opening_bracket(mode);
+    while ((r = mdb_cursor_get(cursor, &key, &value, MDB_NEXT)) == MDB_SUCCESS)
+    {
+        switch (mode) {
+            case DUMP_MODE_DEFAULT:
```

Comment:
I'd call this `DUMP_MODE_FULL` or something like that.

Reply:
Alright, fixed.